File: leetcode/python/649_dota2_senate.py
from collections import deque
from this import d


class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        rCount = senate.count('R')
        dCount = len(senate) - rCount
        bannedR = bannedD = 0
        q = deque(senate)
        while dCount and rCount:
            senator = q.popleft()
            if senator == 'R':
                if bannedR:
                    bannedR -= 1
                    rCount -= 1
                else:
                    bannedD += 1
                    q.append(senator)
            else:
                if bannedD:
                    bannedD -= 1
                    dCount -= 1
                else:
                    bannedR += 1
                    q.append(senator)
        return 'Radiant' if rCount else 'Dire'

File: leetcode/rust/649_dota2_senate.rs
use std::collections::VecDeque;

impl Solution {
    pub fn predict_party_victory(senate: String) -> String {
        let mut q = VecDeque::from(senate.chars().collect::<Vec<char>>());
        let mut banned_d = 0;
        let mut banned_r = 0;  
        let mut r_count = q.iter().filter(|&c| *c == 'R').count();
        let mut d_count = q.len() - r_count;

        while r_count > 0 && d_count > 0 {
            let c = q.pop_front().unwrap();
            if c == 'R' {
                if banned_r > 0 {
                    banned_r -= 1;
                    r_count -= 1
                } else {
                    q.push_back('R');
                    banned_d += 1;
                }
            } else {
                if banned_d > 0 {
                    banned_d -= 1;
                    d_count -= 1;
                } else {
                    banned_r += 1;
                    q.push_back('D');
                }
            }
        }

        if r_count > 0 {
            "Radiant".to_string()
        } else {
            "Dire".to_string()
        }
    }
}